sympy/solvers/tests/test_solvers.py

```diff
@@ -136,15 +137,15 @@ def test_solve_args():
     assert solve(y - 3, {y}) == [3]
     # more than 1
     assert solve(y - 3, {x, y}) == [{y: 3}]
-    # multiple symbols: take the first linear solution+
-    # - return as tuple with values for all requested symbols
-    assert solve(x + y - 3, [x, y]) == [(3 - y, y)]
-    # - unless dict is True
-    assert solve(x + y - 3, [x, y], dict=True) == [{x: 3 - y}]
-    # - or no symbols are given
+    # no symbols are given: find solution for each factor
     assert solve(x + y - 3) == [{x: 3 - y}]
-    # multiple symbols might represent an undetermined coefficients system
+    # multiple symbols: return first solution (linear if possible)
+    assert solve(x + y - 3, [x, y]) == [{x: 3 - y}]
+    # multiple symbols might represent a coefficients system
+    # - that is linear
     assert solve(a + b*x - 2, [a, b]) == {a: 2, b: 0}
+    assert solve((a*x + b*x - b + d), (a, b)) == {a: -d, b: d}
```

Comment:
I don't think that `solve` should do this. This is conceptually very different from what `solve` usually does. These results do not hold for almost all values of `x` e.g.:
```python
In [18]: eq = a + b*x - 2

In [19]: solve(eq, [a, b], dict=True)
Out[19]: [{a: 2, b: 0}]

In [20]: solve(eq.subs(x, 1), [a, b], dict=True)
Out[20]: [{a: 2 - b}]

In [21]: solve(eq.subs(x, 2), [a, b], dict=True)
Out[21]: [{a: 2 - 2⋅b}]
```
For any particular value of `x` there are infinitely many solutions. The correct return here would be:
```python
solve(eq.subs(x, 2), dict=True)
[{a: 2 - x⋅b}]
```

Reply:
Without sentiment of defense, I note that the result in master for your example is the same and it has been like this since 2011 (2ff42e6893812f42147d8cff16dc3f371021ab63). If the user wants to solve for `a` or `b` then they can give a single symbol or not give any symbols and indicate what they don't want (e.g., using `exclude=[x]`).

I actually had in mind comments you have made (but I can't find now) about results which are nearly obviously not what the user would have wanted. (If you can find that thread, it would be helpful).

It might be helpful to separate out the `solve_undetermined_coeffs` from `solve`...but that will be a regression for folks for whom such a request to solve makes sense.

Reply:
I can see that it is already the case on master for that example but I consider that as bug to be fixed and I certainly don't want to add more cases like that.

When the user asks to solve a system of equations with certain symbols given as unknowns any other symbols should be understood as parameters rather than unknowns. It isn't for solve to decide that x=0 if x is not an unknown. We need solve to return a result that is valid generically for almost all possible values of x.

It's quite clear here what that result should be:
```python
In [22]: solve([eq], [a, b])
Out[22]: {a: -b⋅x + 2}

In [23]: linsolve([eq], [a, b])
Out[23]: {(-b⋅x + 2, b)}

In [24]: nonlinsolve([eq], [a, b])
Out[24]: {(-b⋅x + 2, b)}
```
The special case that gives this result should just be removed:
```python
In [25]: solve(eq, [a, b])
Out[25]: {a: 2, b: 0}
```
The equation is underdetermined so solve should not return something that indicates a unique solution.

Reply:
Note that this result is for the one-equation-many-symbols case. If we don't treat that differently than n-equations-more-than-n-symbols then do you propose that we raise an error when more than one symbol is provided? Or just pick an arbitrary variable for which to solve? What's wrong with
```python
eq = f(a,b,x)
solve(eq) -> solve returns a solution for a, b or x
solve(eq, exclude=[x]) or solve([eq],(a,b)) -> solve solves for a or b
solve(eq, (a,b)) -> solve gives values of a AND b that make the equation 0 (case of undetermined coefficients)
```
The case of solving for undetermined coefficients is not an uncommon request. Do you agree that splitting off that behavior is a regression? Any ideas on how to accomplish that?

ping @asmeurer

Reply:
The expected output is this:
```python
In [10]: a, b, x = symbols('a, b, x')

In [11]: eq = a + b*x - 2

In [12]: solve([eq], [a, b])
Out[12]: {a: -b⋅x + 2}
```
This should give the same result:
```python
In [13]: solve(eq, [a, b])
Out[13]: {a: 2, b: 0}
```
There should be no distinction between `eq` and `[eq]` here and the result for `[eq]` is the correct one in this case. It is arbitrary whether `a` should be given in terms of `b` or vice versa in general but in this case it's better to give `a` in terms of `b` because it doesn't require dividing by `x` which might be zero.

Reply:
I would consider fixing it so that `eq` and `[eq]` return the same (correct) results to be a bugfix rather than a regression.

Reply:
cf docstring for solve which explicitly states that one equation with more than one variable will attempt a "undetermined coefficients" solution:
```
Single expression and more than one symbol:
        * When there is a linear solution:
            >>> solve(x - y**2, x, y)
            [(y**2, y)]
            >>> solve(x**2 - y, x, y)
            [(x, x**2)]
            >>> solve(x**2 - y, x, y, dict=True)
            [{y: x**2}]
        * When undetermined coefficients are identified:
            * That are linear:
                >>> solve((a + b)*x - b + 2, a, b)
                {a: -2, b: 2}
            * That are nonlinear:
                >>> solve((a + b)*x - b**2 + 2, a, b, set=True)
                ([a, b], {(-sqrt(2), sqrt(2)), (sqrt(2), -sqrt(2))})
```
Should this be raise on the mailing list before making a change to this?